File: sdk/typescript/src/relays.ts
import type { WlteClient } from './client.js'
import type {
  ApiEnvelope,
  CommandExecution,
  RelayCommandOptions,
  RelayJogConfigOptions,
  RelayJogOptions,
  RelaySetOptions,
} from './types.js'

export class RelaysApi {
  constructor(private readonly client: WlteClient) {}

  async set(deviceId: string, options: RelaySetOptions): Promise<CommandExecution> {
    return this.control(deviceId, {
      relays: [{ index: options.index, action: options.on ? 'ON' : 'OFF' }],
      idempotencyKey: options.idempotencyKey,
    })
  }

  async control(deviceId: string, options: RelayCommandOptions): Promise<CommandExecution> {
    const response = await this.client.request<ApiEnvelope<CommandExecution>>(
      `/wlte/v1/devices/${encodeURIComponent(deviceId)}/relays/commands`,
      {
        method: 'POST',
        headers: {
          'Idempotency-Key': options.idempotencyKey,
        },
        body: {
          relays: options.relays,
        },
      },
    )
    return response.data
  }

  async jog(deviceId: string, options: RelayJogOptions): Promise<CommandExecution> {
    return this.control(deviceId, {
      relays: [{ index: options.index, action: 'JOG' }],
      idempotencyKey: options.idempotencyKey,
    })
  }

  async setJogConfig(deviceId: string, options: RelayJogConfigOptions): Promise<CommandExecution> {
    const response = await this.client.request<ApiEnvelope<CommandExecution>>(
      `/wlte/v1/devices/${encodeURIComponent(deviceId)}/relays/${options.index}/jog-config`,
      {
        method: 'PUT',
        headers: {
          'Idempotency-Key': options.idempotencyKey,
        },
        body: {
          durationSec: options.durationSec,
        },
      },
    )
    return response.data
  }
}

File: sdk/go/relays.go
package wlteopenapi

import (
	"context"
	"fmt"
)

type RelaysService struct {
	client *Client
}

func (s *RelaysService) Set(ctx context.Context, deviceID string, options RelaySetOptions) (CommandExecution, error) {
	return s.Control(ctx, deviceID, RelayCommandOptions{
		Relays:         []RelayCommand{{Index: options.Index, Action: relayActionForBool(options.On)}},
		IdempotencyKey: options.IdempotencyKey,
	})
}

func (s *RelaysService) Control(ctx context.Context, deviceID string, options RelayCommandOptions) (CommandExecution, error) {
	var result CommandExecution
	err := s.client.request(
		ctx,
		"POST",
		fmt.Sprintf("/wlte/v1/devices/%s/relays/commands", urlEscape(deviceID)),
		nil,
		map[string]string{"Idempotency-Key": options.IdempotencyKey},
		map[string][]RelayCommand{"relays": options.Relays},
		&result,
	)
	return result, err
}

func (s *RelaysService) Jog(ctx context.Context, deviceID string, options RelayJogOptions) (CommandExecution, error) {
	return s.Control(ctx, deviceID, RelayCommandOptions{
		Relays:         []RelayCommand{{Index: options.Index, Action: RelayActionJog}},
		IdempotencyKey: options.IdempotencyKey,
	})
}

func (s *RelaysService) SetJogConfig(ctx context.Context, deviceID string, options RelayJogConfigOptions) (CommandExecution, error) {
	var result CommandExecution
	err := s.client.request(
		ctx,
		"PUT",
		fmt.Sprintf("/wlte/v1/devices/%s/relays/%d/jog-config", urlEscape(deviceID), options.Index),
		nil,
		map[string]string{"Idempotency-Key": options.IdempotencyKey},
		map[string]int{"durationSec": options.DurationSec},
		&result,
	)
	return result, err
}

func relayActionForBool(on bool) RelayAction {
	if on {
		return RelayActionOn
	}
	return RelayActionOff
}
